test: port libvterm unit test to neovim test suite

dundargoc opened the pull request:
These were imported from the v0.3.3 git tag
https://github.com/neovim/libvterm/tree/v0.3.3 and not the latest
commit. This is for compatibility reasons as the libvterm code was
imported from v0.3.3.

bfredl commented on src/vterm/vterm.h:
```diff
@@ -499,18 +499,18 @@ void vterm_state_send_selection(VTermState *state, VTermSelectionMask mask, VTer
 // ------------
 
 typedef struct {
-    unsigned int bold      : 1;
```
removing bitfields are going to expand the size of these structs significantly.

If we need to use these in tests it would be better to write parts of the test logic as C helper functions (there is some predecendent for this, marktree e.g. has `ifndef NDEBUG` functions only for debug builds).

bfredl replied:
or if we don't need them we just need to filter out them so luajit doesn't need to see them and get angy..

dundargoc replied:
I suspect they're not needed since both the vterm and our own tests pass without them, but I am not entirely sure. I did it this way for now to unblock the work.

zeertzjq replied:
Did you not read the first sentence?

dundargoc replied:
I did but I don't know how to filter them out like he suggested.

dundargoc replied:
I did, but I misunderstood. I've opted to filter out all bitfields in unit tests.

dundargoc committed:
test: port libvterm unit test to neovim test suite

These were imported from the v0.3.3 git tag
https://github.com/neovim/libvterm/tree/v0.3.3 and not the latest
commit. This is for compatibility reasons as the libvterm code was
imported from v0.3.3.

dundargoc commented on test/unit/testutil.lua:
```diff
@@ -151,6 +151,16 @@ local function filter_complex_blocks(body)
+        return string.format('typedef struct {%s}', s:gsub('%s*:%s*%d+%s*;', ';'))
+      end)
+
+      -- HACK: hardcode to remove difficult bitfield not handled by the previous gsub
+      if line:find('struct VTermState') then
+        line = string.gsub(line, 'state : 8;', 'state;')
+      end
```
Hacky AF, but it works for now

dundargoc commented on src/vterm/vterm.c:
```diff
@@ -429,3 +429,212 @@ void vterm_check_version(int major, int minor)
+
+#define TEST_PARSER_FILE "test_parser_output"
+
+int parser_text(const char bytes[], size_t len, void *user)
+{
+  FILE *f = fopen(TEST_PARSER_FILE, "a");
+  fprintf(f, "text ");
+  int i;
+  for(i = 0; i < len; i++) {
+    unsigned char b = bytes[i];
+    if(b < 0x20 || b == 0x7f || (b >= 0x80 && b < 0xa0)) {
+      break;
+    }
+    fprintf(f, i ? ",%x" : "%x", b);
+  }
+  fprintf(f, "\n");
+  fclose(f);
+
+  return i;
+}
```
TODO: wonder if I can just write this as a lua function instead.

justinmk commented on CMakeLists.txt:
```diff
@@ -46,6 +46,7 @@ set(DEPS_IGNORE_SHA FALSE)
 #-------------------------------------------------------------------------------
 set(FUNCS_DATA ${PROJECT_BINARY_DIR}/funcs_data.mpack)
 set(TOUCHES_DIR ${PROJECT_BINARY_DIR}/touches)
+set(VTERM_TEST_FILE ${PROJECT_BINARY_DIR}/test/vterm_test_output)
```
if this is a test artifact, should this go in the build/ dir? that is where we store e.g. tempfiles created by `t.tmpname()`

dundargoc replied:
`PROJECT_BINARY_DIR` is the `build/` dir.

justinmk commented on src/nvim/terminal.c:
```diff
@@ -232,7 +232,7 @@ static int parse_osc8(VTermStringFragment frag, int *attr)
   // Parse the URI from the OSC 8 sequence and add the URL to our URL set.
   // Skip the ID, we don't use it (for now)
-  size_t i = 0;
+  int i = 0;
```
👀 is this temporary change or is it explained somewhere?

dundargoc replied:
I needed to change the type of `flag.len` to `int`, which causes "Comparison of integers of different signs" if I don't change this to `int`. I can just cast it to `int` every time instead if you prefer.

dundargoc replied:
So not temporary. I can keep it like this or revert to `size_t` and cast `flag.len` to `size_t`, both should work equally well. Lemme know if you have a preference.

justinmk commented on src/nvim/terminal.c:
```diff
@@ -248,21 +248,21 @@ static int parse_osc8(VTermStringFragment frag, int *attr)
   }
 
-  char *url = xmemdupz(&frag.str[start], len + 1);
+  char *url = xmemdupz(&frag.str[start], (size_t)len + 1);
```
might need `assert(len >= 0 && len < MAX_INT32)` for coverity

dundargoc replied:
These pre-emptive coverity warning guessing as well as the post-merged coverity warnings adds a lot of friction. I'll see if I have time to set it up coverity CI check for PRs this week, otherwise we might wanna create a dedicated issue to track this problem.

justinmk commented on src/vterm/vterm.c:
```diff
@@ -429,3 +430,508 @@ void vterm_check_version(int major, int minor)
   // Happy
 }
+
+#ifndef NDEBUG
```
```suggestion
// For unit tests.
#ifndef NDEBUG
```

dundargoc commented on test/unit/vterm_spec.lua:
```diff
@@ -0,0 +1,3590 @@
+    screen_row(0, 'Top', screen)
+  end)
+
+  itp('92lp1640917', function()
```
That is the name of the test file. Don't ask lol.

dundargoc replied:
Shall I change it to something more readable/useful?

justinmk replied:
Just checking that it's intentional. If a more meaningful name is possible then great

zeertzjq commented on src/vterm/vterm.h:
```diff
@@ -219,7 +220,7 @@ typedef enum {
 typedef struct {
   const char *str;
-  size_t      len : 30;
+  int         len : 30;
```
Can you mention somewhere why this change is necessary?

dundargoc replied:
The tests fail with `test/unit/testutil.lua:276: invalid value near ':' at line 18'` otherwise.

dundargoc replied:
Reverted it back to size_t by removing the bitfield in tests.
